File: extensions/libdepwin/nappgui/src/osgui/win/oslistener.c
/*
 * NAppGUI Cross-platform C SDK
 * 2015-2024 Francisco Garcia Collado
 * MIT Licence
 * https://nappgui.com/en/legal/license.html
 *
 * File: oslistener.c
 *
 */

/* View listeners */

#include "oslistener.inl"
#include "oscontrol.inl"
#include "osgui.inl"
#include "osgui_win.inl"
#include "osscrolls.inl"
#include <core/event.h>
#include <osbs/btime.h>
#include <sewer/bmem.h>
#include <sewer/cassert.h>
#include <sewer/types.h>

#if !defined(__WINDOWS__)
#error This file is only for Windows
#endif

#if _MSC_VER > 1400
#define MAPVK_VSC MAPVK_VSC_TO_VK_EX
#else
#define MAPVK_VSC 1
#endif

/*---------------------------------------------------------------------------*/

void oslistener_init(ViewListeners *listeners)
{
    cassert_no_null(listeners);
    bmem_zero(listeners, ViewListeners);
    listeners->enabled = TRUE;
    listeners->button = ENUM_MAX(gui_mouse_t);
}

/*---------------------------------------------------------------------------*/

void oslistener_remove(ViewListeners *listeners)
{
    cassert_no_null(listeners);
    listener_destroy(&listeners->OnDraw);
    listener_destroy(&listeners->OnEnter);
    listener_destroy(&listeners->OnExit);
    listener_destroy(&listeners->OnMoved);
    listener_destroy(&listeners->OnDown);
    listener_destroy(&listeners->OnUp);
    listener_destroy(&listeners->OnClick);
    listener_destroy(&listeners->OnDrag);
    listener_destroy(&listeners->OnWheel);
    listener_destroy(&listeners->OnKeyDown);
    listener_destroy(&listeners->OnKeyUp);
}

/*---------------------------------------------------------------------------*/

void oslistener_set_enabled(ViewListeners *listeners, bool_t enabled)
{
    cassert_no_null(listeners);
    listeners->enabled = enabled;
}

/*---------------------------------------------------------------------------*/

void oslistener_draw(OSControl *sender, DCtx *ctx, const real32_t width, const real32_t height, const real32_t visible_x, const real32_t visible_y, const real32_t visible_width, const real32_t visible_height, ViewListeners *listeners)
{
    cassert_no_null(sender);
    cassert_no_null(listeners);
    if (listeners->OnDraw != NULL)
    {
        EvDraw params;
        params.ctx = ctx;
        params.x = visible_x;
        params.y = visible_y;
        params.width = visible_width;
        params.height = visible_height;
        unref(width);
        unref(height);
        listener_event(listeners->OnDraw, ekGUI_EVENT_DRAW, sender, &params, NULL, OSControl, EvDraw, void);
    }
}

/*---------------------------------------------------------------------------*/

void oslistener_mouse_exit(OSControl *sender, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert(listeners->is_mouse_inside == TRUE);
    listeners->is_mouse_inside = FALSE;
    if (listeners->OnExit != NULL)
        listener_event(listeners->OnExit, ekGUI_EVENT_EXIT, sender, NULL, NULL, OSControl, void, void);
}

/*---------------------------------------------------------------------------*/

void oslistener_mouse_moved(OSControl *sender, WPARAM event_wParam, const real32_t x, const real32_t y, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(sender);
    if (listeners->enabled == TRUE)
    {
        EvMouse params;

        if ((event_wParam & MK_LBUTTON) && listeners->button != ekGUI_MOUSE_LEFT)
            return;

        if ((event_wParam & MK_RBUTTON) && listeners->button != ekGUI_MOUSE_RIGHT)
            return;

        if ((event_wParam & MK_MBUTTON) && listeners->button != ekGUI_MOUSE_MIDDLE)
            return;

        params.x = x + (scroll ? osscrolls_x_pos(scroll) : 0);
        params.y = y + (scroll ? osscrolls_y_pos(scroll) : 0);
        params.lx = x;
        params.ly = y;
        params.button = listeners->button;
        params.count = 0;
        params.modifiers = _osgui_modifiers();
        params.tag = 0;

        if (listeners->button != ENUM_MAX(gui_mouse_t))
        {
            /*if (x >= 0.f && x < frame.size.width && y >= 0.f && y < frame.size.height)*/
            {
                if (listeners->OnDrag != NULL)
                    listener_event(listeners->OnDrag, ekGUI_EVENT_DRAG, sender, &params, NULL, OSControl, EvMouse, void);
            }
        }
        else
        {
            if (listeners->is_mouse_inside == FALSE)
            {
                TRACKMOUSEEVENT track;
                BOOL ok;
                listeners->is_mouse_inside = TRUE;
                track.cbSize = sizeof(TRACKMOUSEEVENT);
                track.dwFlags = /*TME_HOVER | */ TME_LEAVE;
                track.hwndTrack = sender->hwnd;
                track.dwHoverTime = HOVER_DEFAULT;
                ok = TrackMouseEvent(&track);
                cassert_unref(ok == TRUE, ok);

                if (listeners->OnEnter != NULL)
                    listener_event(listeners->OnEnter, ekGUI_EVENT_ENTER, sender, &params, NULL, OSControl, EvMouse, void);
            }
            else if (listeners->OnMoved != NULL)
            {
                listener_event(listeners->OnMoved, ekGUI_EVENT_MOVED, sender, &params, NULL, OSControl, EvMouse, void);
            }
        }
    }
}

/*---------------------------------------------------------------------------*/

void oslistener_mouse_down(OSControl *sender, const gui_mouse_t button, const real32_t x, const real32_t y, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(sender);
    if (listeners->enabled == TRUE)
    {
        listeners->button = button;
        SetCapture(sender->hwnd);

        listeners->mouse_down_x = x;
        listeners->mouse_down_y = y;
        listeners->mouse_down_time = btime_now();

        if (listeners->OnDown != NULL)
        {
            EvMouse params;
            params.x = x + (scroll ? osscrolls_x_pos(scroll) : 0);
            params.y = y + (scroll ? osscrolls_y_pos(scroll) : 0);
            params.lx = x;
            params.ly = y;
            params.button = button;
            params.count = 0;
            params.modifiers = _osgui_modifiers();
            params.tag = 0;
            listener_event(listeners->OnDown, ekGUI_EVENT_DOWN, sender, &params, NULL, OSControl, EvMouse, void);
        }
    }
}

/*---------------------------------------------------------------------------*/

void oslistener_mouse_up(OSControl *sender, const gui_mouse_t button, const real32_t x, const real32_t y, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(sender);
    if (listeners->enabled == TRUE)
    {
        EvMouse params;
        params.x = x + (scroll ? osscrolls_x_pos(scroll) : 0);
        params.y = y + (scroll ? osscrolls_y_pos(scroll) : 0);
        params.lx = x;
        params.ly = y;
        params.button = button;
        params.count = 0;
        params.modifiers = _osgui_modifiers();
        params.tag = 0;

        {
            BOOL ok = ReleaseCapture();
            cassert_unref(ok != 0, ok);
        }

        if (listeners->OnUp != NULL)
            listener_event(listeners->OnUp, ekGUI_EVENT_UP, sender, &params, NULL, OSControl, EvMouse, void);

        if (listeners->OnClick != NULL)
        {
            // Milliseconds
            uint64_t ellapsed = (btime_now() - listeners->mouse_down_time) / 1000;
            UINT max = GetDoubleClickTime();
            real32_t dx = abs_r32(x - listeners->mouse_down_x);
            real32_t dy = abs_r32(y - listeners->mouse_down_y);
            if (ellapsed < max && dx < 5 && dy < 5)
                listener_event(listeners->OnClick, ekGUI_EVENT_CLICK, sender, &params, NULL, OSControl, EvMouse, void);
        }

        listeners->button = ENUM_MAX(gui_mouse_t);
    }
}

/*---------------------------------------------------------------------------*/

void oslistener_whell(OSControl *sender, WPARAM event_wParam, LPARAM event_lParam, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(sender);
    if (listeners->enabled == TRUE)
    {
        if (listeners->OnWheel != NULL)
        {
            POINTS spoint;
            POINT point;
            BOOL ok;
            EvWheel params;
            spoint = MAKEPOINTS(event_lParam);
            point.x = spoint.x;
            point.y = spoint.y;
            ok = ScreenToClient(sender->hwnd, &point);
            cassert_unref(ok == TRUE, ok);
            params.x = (real32_t)point.x + (scroll ? osscrolls_x_pos(scroll) : 0);
            params.y = (real32_t)point.y + (scroll ? osscrolls_y_pos(scroll) : 0);
            params.dx = 0;
            params.dy = (real32_t)(GET_WHEEL_DELTA_WPARAM(event_wParam) / WHEEL_DELTA);
            params.dz = 0;
            listener_event(listeners->OnWheel, ekGUI_EVENT_WHEEL, sender, &params, NULL, OSControl, EvWheel, void);
        }
    }
}

/*---------------------------------------------------------------------------*/

static bool_t i_key_event(OSControl *sender, const uint32_t type, WPARAM wParam, LPARAM lParam, Listener *listener)
{
    cassert_no_null(sender);
    cassert_unref(sender->type == ekGUI_TYPE_CUSTOMVIEW, sender);
    if (listener != NULL)
    {
        vkey_t key = ENUM_MAX(vkey_t);

        if (wParam == VK_SHIFT)
        {
            register UINT scancode = (lParam & 0x00ff0000) >> 16;
            wParam = MapVirtualKey(scancode, MAPVK_VSC);
            if (wParam == VK_LSHIFT)
                key = ekKEY_LSHIFT;
            else if (wParam == VK_RSHIFT)
                key = ekKEY_RSHIFT;
        }
        else if (wParam == VK_CONTROL)
        {
            if ((lParam & 0x01000000) != 0)
                key = ekKEY_RCTRL;
            else
                key = ekKEY_LCTRL;
        }
        else if (wParam == VK_MENU)
        {
            if ((lParam & 0x01000000) != 0)
                key = ekKEY_RALT;
            else
                key = ekKEY_LALT;
        }
        else
        {
            key = _osgui_vkey((WORD)wParam);
        }

        if (key != ENUM_MAX(vkey_t))
        {
            EvKey params;
            params.key = key;
            params.modifiers = _osgui_modifiers();
            listener_event(listener, type, sender, &params, NULL, OSControl, EvKey, void);
            return TRUE;
        }
    }

    return FALSE;
}

/*---------------------------------------------------------------------------*/

bool_t oslistener_key_down(OSControl *sender, WPARAM wParam, LPARAM lParam, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    return i_key_event(sender, ekGUI_EVENT_KEYDOWN, wParam, lParam, listeners->OnKeyDown);
}

/*---------------------------------------------------------------------------*/

bool_t oslistener_key_up(OSControl *sender, WPARAM wParam, LPARAM lParam, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    return i_key_event(sender, ekGUI_EVENT_KEYUP, wParam, lParam, listeners->OnKeyUp);
}

File: extensions/libdepwin/nappgui/src/osgui/osx/oslistener.m
/*
 * NAppGUI Cross-platform C SDK
 * 2015-2024 Francisco Garcia Collado
 * MIT Licence
 * https://nappgui.com/en/legal/license.html
 *
 * File: oslistener.m
 *
 */

/* View listeners */

#include "oslistener.inl"
#include "osgui.inl"
#include "osscrolls.inl"
#include "osgui_osx.inl"
#include <core/event.h>
#include <sewer/bmem.h>
#include <sewer/cassert.h>

#if !defined(__MACOS__)
#error This file is only for OSX
#endif

/*---------------------------------------------------------------------------*/

void _oslistener_init(ViewListeners *listeners)
{
    cassert_no_null(listeners);
    bmem_zero(listeners, ViewListeners);
    /*listeners->is_dirty = YES;*/
    listeners->is_enabled = YES;
}

/*---------------------------------------------------------------------------*/

void _oslistener_release(ViewListeners *listeners)
{
    cassert_no_null(listeners);
    listener_destroy(&listeners->OnDraw);
    listener_destroy(&listeners->OnEnter);
    listener_destroy(&listeners->OnExit);
    listener_destroy(&listeners->OnMoved);
    listener_destroy(&listeners->OnDown);
    listener_destroy(&listeners->OnUp);
    listener_destroy(&listeners->OnClick);
    listener_destroy(&listeners->OnDrag);
    listener_destroy(&listeners->OnWheel);
    listener_destroy(&listeners->OnKeyUp);
    listener_destroy(&listeners->OnKeyDown);
}

/*---------------------------------------------------------------------------*/

void _oslistener_set_enabled(ViewListeners *listeners, bool_t is_enabled)
{
    cassert_no_null(listeners);
    listeners->is_enabled = (is_enabled == TRUE) ? YES : NO;
}

/*---------------------------------------------------------------------------*/

static void i_mouse_position_in_view_coordinates(const NSView *view, const NSPoint mouse_location_in_window, real32_t *x, real32_t *y)
{
    NSPoint local_point;
    cassert_no_null(view);
    cassert_no_null(x);
    cassert_no_null(y);
    local_point = [view convertPoint:mouse_location_in_window fromView:nil];
    *x = (real32_t)local_point.x;
    *y = (real32_t)local_point.y;
    /* Note: The y coordinate in the returned point starts from a base of 1, not 0. */
    /* *y -= 1.f; */
    /* *y = (real32_t)[view frame].size.height - *y - 1.f; */
}

/*---------------------------------------------------------------------------*/

void _oslistener_mouse_position_in_view_coordinates(const NSView *view, const NSPoint mouse_location_in_window, real32_t *x, real32_t *y)
{
    i_mouse_position_in_view_coordinates(view, mouse_location_in_window, x, y);
}

/*---------------------------------------------------------------------------*/

void _oslistener_mouse_enter(const NSView *view, NSEvent *theEvent, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(theEvent);
    if (listeners->is_enabled && listeners->OnEnter != NULL)
    {
        EvMouse params;
        i_mouse_position_in_view_coordinates(view, [theEvent locationInWindow], &params.lx, &params.ly);
        params.button = ENUM_MAX(gui_mouse_t);
        params.count = 0;
        params.x = params.lx;
        params.y = params.ly;
        params.modifiers = osgui_modifiers([theEvent modifierFlags]);
        params.tag = 0;

        if (scroll != NULL)
        {
            params.x += osscrolls_x_pos(scroll);
            params.y += osscrolls_y_pos(scroll);
        }
        listener_event(listeners->OnEnter, ekGUI_EVENT_ENTER, (OSView *)view, &params, NULL, OSView, EvMouse, void);
    }
}

/*---------------------------------------------------------------------------*/

void _oslistener_mouse_exit(const NSView *view, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    if (listeners->is_enabled && listeners->OnExit != NULL)
        listener_event(listeners->OnExit, ekGUI_EVENT_EXIT, (OSView *)view, NULL, NULL, OSView, void, void);
}

/*---------------------------------------------------------------------------*/

void _oslistener_mouse_moved(const NSView *view, NSEvent *theEvent, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(theEvent);
    if (listeners->is_enabled && listeners->OnMoved != NULL)
    {
        real32_t x, y;
        /*NSSize size = [view frame].size;*/
        i_mouse_position_in_view_coordinates(view, [theEvent locationInWindow], &x, &y);
        /*if (x >= 0.f && x < size.width && y >= 0.f && y < size.height)*/
        {
            EvMouse params;
            params.lx = x;
            params.ly = y;
            params.button = ENUM_MAX(gui_mouse_t);
            params.count = 0;
            params.x = params.lx;
            params.y = params.ly;
            params.modifiers = osgui_modifiers([theEvent modifierFlags]);
            params.tag = 0;

            if (scroll != NULL)
            {
                params.x += osscrolls_x_pos(scroll);
                params.y += osscrolls_y_pos(scroll);
            }
            listener_event(listeners->OnMoved, ekGUI_EVENT_MOVED, (OSView *)view, &params, NULL, OSView, EvMouse, void);
        }
    }
}

/*---------------------------------------------------------------------------*/

void _oslistener_mouse_down(const NSView *view, NSEvent *theEvent, const gui_mouse_t button, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(theEvent);
    if (listeners->is_enabled && listeners->OnDown != NULL)
    {
        EvMouse params;
        i_mouse_position_in_view_coordinates(view, [theEvent locationInWindow], &params.lx, &params.ly);
        params.button = button;
        params.count = 0;
        params.x = params.lx;
        params.y = params.ly;
        params.modifiers = osgui_modifiers([theEvent modifierFlags]);
        params.tag = 0;

        if (scroll != NULL)
        {
            params.x += osscrolls_x_pos(scroll);
            params.y += osscrolls_y_pos(scroll);
        }
        listener_event(listeners->OnDown, ekGUI_EVENT_DOWN, (OSView *)view, &params, NULL, OSView, EvMouse, void);
    }
}

/*---------------------------------------------------------------------------*/

void _oslistener_mouse_up(const NSView *view, NSEvent *theEvent, const gui_mouse_t button, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(theEvent);
    if (listeners->is_enabled == YES)
    {
        real32_t x, y;
        cassert_no_null(theEvent);
        i_mouse_position_in_view_coordinates(view, [theEvent locationInWindow], &x, &y);

        if (listeners->OnUp != NULL)
        {
            EvMouse params;
            params.lx = x;
            params.ly = y;
            params.button = button;
            params.count = 0;
            params.x = params.lx;
            params.y = params.ly;
            params.modifiers = osgui_modifiers([theEvent modifierFlags]);
            params.tag = 0;

            if (scroll != NULL)
            {
                params.x += osscrolls_x_pos(scroll);
                params.y += osscrolls_y_pos(scroll);
            }
            listener_event(listeners->OnUp, ekGUI_EVENT_UP, (OSView *)view, &params, NULL, OSView, EvMouse, void);
        }

        if (listeners->OnClick != NULL)
        {
            EvMouse params;
            params.x = x;
            params.y = y;
            params.button = button;
            params.count = (uint32_t)[theEvent clickCount];
            params.modifiers = osgui_modifiers([theEvent modifierFlags]);
            params.tag = 0;
            listener_event(listeners->OnClick, ekGUI_EVENT_CLICK, (OSView *)view, &params, NULL, OSView, EvMouse, void);
        }
    }
}

/*---------------------------------------------------------------------------*/

void _oslistener_mouse_dragged2(const NSView *view, NSEvent *theEvent, const gui_mouse_t button, const OSScrolls *scroll, Listener *OnDrag_listener)
{
    cassert_no_null(theEvent);
    if (OnDrag_listener != NULL)
    {
        EvMouse params;
        i_mouse_position_in_view_coordinates(view, [theEvent locationInWindow], &params.lx, &params.ly);
        params.button = button;
        params.count = 0;
        params.x = params.lx;
        params.y = params.ly;
        params.modifiers = osgui_modifiers([theEvent modifierFlags]);
        params.tag = 0;

        if (scroll != NULL)
        {
            params.x += osscrolls_x_pos(scroll);
            params.y += osscrolls_y_pos(scroll);
        }
        listener_event(OnDrag_listener, ekGUI_EVENT_DRAG, (OSView *)view, &params, NULL, OSView, EvMouse, void);
    }
}

/*---------------------------------------------------------------------------*/

void _oslistener_mouse_dragged(const NSView *view, NSEvent *theEvent, const gui_mouse_t button, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    if (listeners->is_enabled == YES)
        _oslistener_mouse_dragged2(view, theEvent, button, scroll, listeners->OnDrag);
}

/*---------------------------------------------------------------------------*/

void _oslistener_scroll_whell(const NSView *view, NSEvent *theEvent, const OSScrolls *scroll, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    if (listeners->OnWheel != NULL && listeners->is_enabled == YES)
    {
        EvWheel params;
        cassert_no_null(theEvent);
        i_mouse_position_in_view_coordinates(view, [theEvent locationInWindow], &params.x, &params.y);
        params.dx = (real32_t)[theEvent deltaX];
        params.dy = (real32_t)[theEvent deltaY];
        params.dz = (real32_t)[theEvent deltaZ];
        if (scroll != NULL)
        {
            params.x += osscrolls_x_pos(scroll);
            params.y += osscrolls_y_pos(scroll);
        }
        listener_event(listeners->OnWheel, ekGUI_EVENT_WHEEL, (OSView *)view, &params, NULL, OSView, EvWheel, void);
    }
}

/*---------------------------------------------------------------------------*/

static __INLINE void i_launch_key_event(const NSView *view, const gui_event_t evtype, vkey_t virtual_key_code, const uint32_t modifiers, Listener *OnKey)
{
    EvKey params;
    cassert_no_null(OnKey);
    params.key = virtual_key_code;
    params.modifiers = modifiers;
    listener_event(OnKey, evtype, (OSView *)view, &params, NULL, OSView, EvKey, void);
}

/*---------------------------------------------------------------------------*/

static void i_process_key_event(const NSView *view, NSEvent *theEvent, const gui_event_t evtype, Listener *OnKey)
{
    vkey_t vkey = osgui_vkey([theEvent keyCode]);
    if (vkey != ENUM_MAX(vkey_t))
    {
        uint32_t modifiers = osgui_modifiers((NSUInteger)[theEvent modifierFlags]);
        i_launch_key_event(view, evtype, vkey, modifiers, OnKey);
    }
}

/*---------------------------------------------------------------------------*/

void _oslistener_key_down(const NSView *view, NSEvent *theEvent, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    if (listeners->is_enabled == YES && listeners->OnKeyDown != NULL)
        i_process_key_event(view, theEvent, ekGUI_EVENT_KEYDOWN, listeners->OnKeyDown);
}

/*---------------------------------------------------------------------------*/

void _oslistener_key_up(const NSView *view, NSEvent *theEvent, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    if (listeners->is_enabled == YES && listeners->OnKeyUp != NULL)
        i_process_key_event(view, theEvent, ekGUI_EVENT_KEYUP, listeners->OnKeyUp);
}

/*---------------------------------------------------------------------------*/

static __INLINE void i_flags_event(const NSView *view, NSUInteger flags, const bool_t press, const bool_t prevpress, const vkey_t key, Listener *OnKeyDown, Listener *OnKeyUp)
{
    if (press != prevpress)
    {
        if (press == TRUE)
        {
            if (OnKeyDown != NULL)
            {
                uint32_t modifiers = osgui_modifiers(flags);
                i_launch_key_event(view, ekGUI_EVENT_KEYDOWN, key, modifiers, OnKeyDown);
            }
        }
        else
        {
            if (OnKeyUp != NULL)
            {
                uint32_t modifiers = osgui_modifiers(flags);
                i_launch_key_event(view, ekGUI_EVENT_KEYUP, key, modifiers, OnKeyUp);
            }
        }
    }
}

/*---------------------------------------------------------------------------*/

void _oslistener_key_flags_changed(const NSView *view, NSEvent *theEvent, ViewListeners *listeners)
{
    cassert_no_null(listeners);
    cassert_no_null(theEvent);
    if (listeners->is_enabled == YES)
    {
        NSUInteger flags;
        bool_t rshift, rctrl, rcommand, ralt;
        bool_t lshift, lctrl, lcommand, lalt;
        bool_t prshift, prctrl, prcommand, pralt;
        bool_t plshift, plctrl, plcommand, plalt;
        flags = [theEvent modifierFlags];
        osgui_modifier_flags(flags, &rshift, &rctrl, &rcommand, &ralt, &lshift, &lctrl, &lcommand, &lalt);
        osgui_modifier_flags(listeners->modifier_flags, &prshift, &prctrl, &prcommand, &pralt, &plshift, &plctrl, &plcommand, &plalt);
        i_flags_event(view, flags, rshift, prshift, ekKEY_RSHIFT, listeners->OnKeyDown, listeners->OnKeyUp);
        i_flags_event(view, flags, lshift, plshift, ekKEY_LSHIFT, listeners->OnKeyDown, listeners->OnKeyUp);
        i_flags_event(view, flags, rctrl, prctrl, ekKEY_RCTRL, listeners->OnKeyDown, listeners->OnKeyUp);
        i_flags_event(view, flags, lctrl, plctrl, ekKEY_LCTRL, listeners->OnKeyDown, listeners->OnKeyUp);
        i_flags_event(view, flags, rcommand, prcommand, ekKEY_RWIN, listeners->OnKeyDown, listeners->OnKeyUp);
        i_flags_event(view, flags, lcommand, plcommand, ekKEY_LWIN, listeners->OnKeyDown, listeners->OnKeyUp);
        i_flags_event(view, flags, ralt, pralt, ekKEY_RALT, listeners->OnKeyDown, listeners->OnKeyUp);
        i_flags_event(view, flags, lalt, plalt, ekKEY_LALT, listeners->OnKeyDown, listeners->OnKeyUp);
        listeners->modifier_flags = flags;
    }
}
